src/Orleans.Runtime/Core/Dispatcher.cs

```diff
@@ -215,238 +68,77 @@ public void ReceiveMessage(Message message)
-        {
-            var overloadException = targetActivation.CheckOverloaded(logger);
-            if (overloadException != null)
-            {
-                MessagingProcessingStatisticsGroup.OnDispatcherMessageProcessedError(message);
-                RejectMessage(message, Message.RejectionTypes.Overloaded, overloadException, "Target activation is overloaded " + targetActivation);
-                return;
-            }
-
-            switch (targetActivation.EnqueueMessage(message))
-            {
-                case ActivationData.EnqueueMessageResult.Success:
-                    // Great, nothing to do
-                    break;
-                case ActivationData.EnqueueMessageResult.ErrorInvalidActivation:
-                    ProcessRequestToInvalidActivation(message, targetActivation.Address, targetActivation.ForwardingAddress, "EnqueueRequest");
-                    break;
-                case ActivationData.EnqueueMessageResult.ErrorActivateFailed:
-                    ProcessRequestToInvalidActivation(message, targetActivation.Address, targetActivation.ForwardingAddress, "EnqueueRequest", rejectMessages: true);
-                    break;
-                case ActivationData.EnqueueMessageResult.ErrorStuckActivation:
-                    // Avoid any new call to this activation
-                    ProcessRequestToStuckActivation(message, targetActivation, "EnqueueRequest - blocked grain");
-                    break;
-                default:
-                    throw new ArgumentOutOfRangeException();
-            }
-
-            // Dont count this as end of processing. The message will come back after queueing via HandleIncomingRequest.
-
-#if DEBUG
-            // This is a hot code path, so using #if to remove diags from Release version
-            // Note: Caller already holds lock on activation
-            if (logger.IsEnabled(LogLevel.Trace)) logger.Trace(ErrorCode.Dispatcher_EnqueueMessage,
-                "EnqueueMessage for {0}: targetActivation={1}", message.TargetActivation, targetActivation.DumpStatus());
-#endif
+                },
+                catalog);
         }
 
         internal void ProcessRequestToInvalidActivation(
-            Message message, 
-            ActivationAddress oldAddress, 
-            ActivationAddress forwardingAddress, 
+            Message message,
+            ActivationAddress oldAddress,
+            ActivationAddress forwardingAddress,
             string failedOperation,
             Exception exc = null,
             bool rejectMessages = false)
         {
             // Just use this opportunity to invalidate local Cache Entry as well. 
             if (oldAddress != null)
             {
-                this.localGrainDirectory.InvalidateCacheEntry(oldAddress);
+                var localGrainDirectory = this.RuntimeClient.ServiceProvider.GetRequiredService<ILocalGrainDirectory>();
```

Comment:
Any reason to not have it in the constructor anymore? To refactor in the future more easily?

Reply:
No good reason, at least not anymore. It probably made sense to me at the time (maybe to break a cyclic loop which no longer exists).

Thanks